src/gen9_avc_encoder.c

```diff
@@ -3926,55 +4442,378 @@ gen9_avc_send_surface_mbenc(VADriverContextP ctx,
-    int mad_enable = 0;
-    VAEncSliceParameterBufferH264 * slice_param = avc_state->slice_param[0];
+    frame_mb_nums = generic_state->frame_width_in_mbs *
+                    generic_state->frame_height_in_mbs;
+    fei_param = avc_state->fei_framectl_param;
+
+    assert(fei_param != NULL);
+
+    /* res_mb_code_surface for MB code */
+    size = frame_mb_nums * FEI_AVC_MB_CODE_BUFFER_SIZE;
+    if (fei_param->mb_code_data != VA_INVALID_ID) {
+        obj_buffer = BUFFER(fei_param->mb_code_data);
+        buffer_store = obj_buffer->buffer_store;
+        assert(size <= buffer_store->bo->size);
+        if (avc_priv_surface->res_mb_code_surface.bo != NULL)
+            i965_free_gpe_resource(&avc_priv_surface->res_mb_code_surface);
+        i965_dri_object_to_buffer_gpe_resource(
+            &avc_priv_surface->res_mb_code_surface,
+            buffer_store->bo);
+    } else {
+        allocate_flag = i965_allocate_gpe_resource(i965->intel.bufmgr,
+                                                   &avc_priv_surface->res_mb_code_surface,
+                                                   ALIGN(size, 0x1000),
+                                                   "mb code buffer");
+        assert(allocate_flag != 0);
+    }
+
+    /* res_mv_data_surface for MV data */
+    size = frame_mb_nums * FEI_AVC_MV_DATA_BUFFER_SIZE;
+    if (fei_param->mv_data != VA_INVALID_ID) {
+        obj_buffer = BUFFER(fei_param->mv_data);
+        buffer_store = obj_buffer->buffer_store;
+        assert(size <= buffer_store->bo->size);
+        if (avc_priv_surface->res_mv_data_surface.bo != NULL)
+            i965_free_gpe_resource(&avc_priv_surface->res_mv_data_surface);
+        i965_dri_object_to_buffer_gpe_resource(
+            &avc_priv_surface->res_mv_data_surface,
+            buffer_store->bo);
+    } else {
+        allocate_flag = i965_allocate_gpe_resource(i965->intel.bufmgr,
+                                                   &avc_priv_surface->res_mv_data_surface,
+                                                   ALIGN(size, 0x1000),
+                                                   "mv data buffer");
+        assert(allocate_flag != 0);
+    }
+
+    /* fei mb control data surface */
+    size = frame_mb_nums * FEI_AVC_MB_CONTROL_BUFFER_SIZE;
+    if ((fei_param->mb_input | fei_param->mb_size_ctrl) &&
+        fei_param->mb_ctrl != VA_INVALID_ID) {
+
+        obj_buffer = BUFFER(fei_param->mb_ctrl);
+        buffer_store = obj_buffer->buffer_store;
+        assert(size <= buffer_store->bo->size);
+        if (avc_priv_surface->res_fei_mb_cntrl_surface.bo != NULL)
+            i965_free_gpe_resource(&avc_priv_surface->res_fei_mb_cntrl_surface);
+        i965_dri_object_to_buffer_gpe_resource(
+            &avc_priv_surface->res_fei_mb_cntrl_surface,
+            buffer_store->bo);
+
+    } else if (fei_param->mb_input | fei_param->mb_size_ctrl) {
+        assert(fei_param->mb_ctrl != VA_INVALID_ID);
+    }
```

Comment:
Would better change it to: 
 if ((fei_param->mb_input | fei_param->mb_size_ctrl) {
 assert(fei_param->mb_ctrl != VA_INVALID_ID);
obj_buffer = BUFFER(fei_param->mb_ctrl);
...
}